### ISSUE TITLE
Bevy Standard Widgets: Roadmap

### ISSUE THREAD

**Issue by viridia** [bevyengine/bevy#19236]
This is a tracking issue for the work described in discussion #16900 and referencing the prototype implementation which can be found here: https://github.com/viridia/bevy_core_widgets

* [x] Upstream interaction state components (`Checked`, `Hovering` and friends)
* [ ] Core widgets, with examples
  * [x] Button
  * [x] Slider
  * [x] Checkbox
  * [x] Radio Group and Radio Button
  * [x] Scrollbar
  * [ ] Number SpinBox (edit numeric fields by dragging or keyboard)
  * [ ] Discrete SpinBox (edit discrete choices by dragging or keyboard)
  * [x] Menu Button
  * [ ] Dialog
  * [x] Text Input
  * [ ] Notification
* [x] Upstream cursor management framework
* [x] Implement theming
* [ ] Implement opinionated editor widgets, with examples
  * [x] Settle on final naming for opinionated widget collection
  * [x] Button
  * [x] Slider
  * [ ] Combo Slider (Slider with editable text)
  * [x] Checkbox
  * [x] Toggle Switch (using Core Checkbox)
  * [ ] Disclosure Triangle (prototype done)
  * [x] Radio Group and Radio Button
  * [ ] Scrollbar and ScrollArea (prototype done)
  * [ ] List View (prototype done)
  * [ ] Combo Number Spinner
  * [x] Menu Button (prototype done)
  * [ ] Dialog
  * [x] Text Input
  * [x] Color Gradient Slider
  * [x] Color Plane (2-axis picker)
  * [x] Color Swatch with Alpha Grid
  * [ ] Swatch Grid
  * [ ] Tab Deck
  * [ ] Tree View
* [ ] Implement opinionated editor layout components (post-BSN)
  * [ ] Page layout variants
  * [ ] Toolbar
  * [ ] Sidebar

Spin-off tasks:
* [x] Rounding options for sliders and other numeric inputs
* [ ] UI Transition animations framework: switch slider, disclosure triangle rotate, dialog enter/exit
* [x] #20499
* [ ] Verify that variable-width fonts work in Bevy, and then replace FiraSans.
* [x] Focus Rings
* [ ] #20524
* [x] Add Reflection support to CoreWidgets

@alice-i-cecile

**Comment by alice-i-cecile** [comment]
I think we should be aiming to build a first-party inspector as our first major stress test / tool for this. This shapes my opinions below.

Prioritize:

- text input
- context menu
- dialog
- discrete spinner (needed to select enums)
- scrollbar (useful for inspector, useful for games)
- tab deck (very useful for complex tools)
- tree view (unblocks first-party inspector)
- pane docking (central UI paradigm for editor)
- disclosure triangle (helpful for inspector)

Deprioritize:

- toggle switch (can use checkboxes)
- number spinner (can use slider or text input)
- swatch grid (needed much later)
- color gradient slider (color picker is more important, but even that isn't pressing)

Clarity needed:

- menu bar: do you mean hamburger menu (unimportant) or top-of-screen menu (important but later)

**Comment by viridia** [comment]
@alice-i-cecile Answers to your questions:

**Menus**: there are several kinds of menus we can do:

* **MenuButton** - this is a button which when clicked displays a drop-down menu. The menu contains actions.
* **Select** - this is like a MenuButton, except that the menu items contain values, not actions. The label of the button changes to the selected value. (The naming is unfortunate, but this is what it's generally called.)
* **MenuBar** - this is a top-of-screen menu
* **Context Menu** - like a right-click menu, it appears at the cursor position. I'm not exactly sure how to approach this, as there are a bunch of different ways we could implement the positioning.
* **Autocomplete** / **Autosuggest** - this is a text input field with a drop-down menu that offers suggestions as you type. Note: this is a very advanced widget, with many potential variations.

**Discrete Spinner**: this is a widget that is not commonly seen in desktop UIs but is frequently see in console settings UIs. My reason for including it in the core widgets collection is that I wanted to do at least one console-centric widget.

(Note that "spinner" may not be the correct term here. I'm referring to a widget that allows cycling through a set of choices, either by dragging or by button input, not the "busy wait spinner" popular in web sites).

Some widgets that I did not include, mainly to keep the list size manageable:

* Linear progress
* Circular progress
* Tooltip
* Knob - like a slider, but rotates. Frequently seen in audio apps.

In the world of React UI libraries, there are a lot of components that we probably won't use:

* Breadcrumbs
* Card
* Avatar
* Badge (a small indicator attached to another widget - like the number of unread notifications)
* Chip - an oval-shaped tag, github labels are an example
* Snackbar - a transient popup shown after another action, like "Undo" after deleting mail
* Drawer

**Comment by jordanhalase** [comment]
I have a suggestion for the number input, feel free to chime in.

Could it use fixed point for its internal representation? Even if only as an option? For example, 2.718 could be represented as some `fixed_value: i64 = 2718` and `fixed_scale: u64 = 3`. It would make fine point stepping lossless and not susceptible to floating point precision issues, and could implement a `get_f32_value()` so users could extract an `f32` or `f64` out of the widget converted from its internal fixed value for use. (As well as allowing the user to extract the raw integer data itself)

There could be a second variant of the number input that uses floats internally for data that actually doesn't work well with fixed point. (Ability to handle microscopic and astronomical data in the same format)

The ability to handle hexadecimal would also be useful for special use cases. The fixed point representation could also have some `fixed_radix` set to 10 for decimal or 16 for hex inputs here.

**Comment by lomirus** [comment]
@jordanhalase I believe implementing a dedicated `get_f32_value()` method for generic input fields is overly specialized. This approach isn't flexible enough to handle various complex requirements in real-world scenarios. I propose decoupling the internal data representation from the display value and exposing two interfaces: `into` and `from`:

```rust
struct TextInput<T, F, I>
where
    F: Fn(String) -> T,
    I: Fn(T) -> String,
{
    value: String,
    inner_value: T,
    from: F,
    into: I,
}
```

```rust
TextInput {
    value: String::new(),
    inner_value: 0,
    from: |s| s.parse().unwrap(),
    into: |i| i.to_string(),
}
```

This design would allow us to:
- Easily implement scenarios where values are stored internally as integers while displaying them as floating-point numbers
- Allow users to freely choose to use any type such as `i32`, `f64`, or `BigInt`.
- Use `from` and `into` to restrict user input to a specific format. e.g., integers only, where the `into` implementation would prevent decimal points from appearing in the `value` field

The code above is purely illustrative of course. We could potentially leverage Rust's `From` and `Into` traits to make it more in line with the Rust paradigm.

**Comment by viridia** [comment]
Before the web had dedicated numeric input fields, the way this was done is to wrap a normal text input inside a composite widget with a custom validator hook. With the right validation callback, you can do integers, floats, phone numbers, email addresses, or whatever specialized formats you want.

**Comment by viridia** [comment]
Here's a progress update:

Menus/MenuButtons/Popups are basically done, just waiting on BSN merge.

Most of the other widgets are not started. The reason for this is because the addition of BSN changes the API a lot, and I don't see much point in giving users an API and then immediately changing it. For example, the disclosure triangle is the same as a toggle switch, but with a different appearance, implementing it is just cut & paste. Other than the BSN dependency, most widgets are fairly easy to do.

Dialogs are partially done.

Color sliders are awaiting on fixes to HSL gradients.

Color swatch is checked in, but will likely be re-implemented a different way in the future.

UI transition framework: this is contentious; I prototyped something, but the reaction on discord was "you're just duplicating work that other people have already done in third-party crates". That's fine, but until those crates are upstreamed, or Bevy gets equivalent functionality, we can't use them in feathers.

Better callbacks: I think the way forward here is garbage-collectible systems, but this isn't something I can implement myself.

Cursors for disabled widgets: This is a super-easy task, I was saving it as a starter task for new Bevy contributors.

Focus Rings: Contention around whether to use outlines or some other method of highlighting.

AccessibleName component: This will work a lot better with BSN.

Add Reflection support to CoreWidgets: I'm not actually sure how to do this, it's more complex than it seems at first glance.

**Comment by alice-i-cecile** [comment]
Text input will likely be solved via #19752.

**Comment by lomirus** [comment]
Could we consider adding a `Notification` core wdiget? This type of widget is also frequently seen in many games—for example, when a player gains or loses an item, levels up, etc., a notification appears to alert the user. Typically, this widget disappears automatically after a certain amount of time. Considering how common this is, I believe it's worthwhile to add a generic widget for this.

**Comment by nic96** [comment]
I love the number input in Blender; if the input has a min and max it visually fills like a slider. The min and max shouldn't necessarily be a hard limit (maybe configurable), but it could help in computing "sensitivity" when dragging on it. Additionally if a modifier is pressed like 'Shift ' it further reduces the sensitivity when dragging. Maybe the SpinBox could work similarly even if it's not directly editable like a number input initially.
